File: DouyinBarrageGrab/Demos/Python/parsers.py
from typing import Dict, Any, Optional
from entities import (
    MsgUser, FansClubInfo, Msg, LikeMsg, MemberMessage, UserSeqMsg, 
    Gender, GiftMsg, FansclubMsg, ShareMessage, RoomAnchorInfo, ShareType
)

# 消息解析器
class MessageParser:
    @staticmethod
    def parse_user(user_data: Dict[str, Any]) -> MsgUser:
        """解析用户数据"""
        user = MsgUser()
        if not user_data:
            return user
            
        user.Id = user_data.get("Id", 0)
        user.IsAdmin = user_data.get("IsAdmin", False)
        user.IsAnchor = user_data.get("IsAnchor", False)
        user.ShortId = user_data.get("ShortId", 0)
        user.DisplayId = user_data.get("DisplayId", "")
        user.Nickname = user_data.get("Nickname", "匿名用户")
        user.Level = user_data.get("Level", 0)
        user.PayLevel = user_data.get("PayLevel", 0)
        user.Gender = Gender(user_data.get("Gender", 0))
        user.HeadImgUrl = user_data.get("HeadImgUrl", "")
        user.SecUid = user_data.get("SecUid", "")
        user.FollowerCount = user_data.get("FollowerCount", 0)
        user.FollowStatus = user_data.get("FollowStatus", 0)
        user.FollowingCount = user_data.get("FollowingCount", 0)
        
        # 解析粉丝团信息
        if "FansClub" in user_data and user_data["FansClub"]:
            fans_club_data = user_data["FansClub"]
            user.FansClub = FansClubInfo(
                ClubName=fans_club_data.get("ClubName", ""),
                Level=fans_club_data.get("Level", 0)
            )
            
        return user
    
    @staticmethod
    def parse_anchor_info(owner_data: Dict[str, Any]) -> Optional[RoomAnchorInfo]:
        """解析主播信息"""
        if not owner_data:
            return None
            
        owner = RoomAnchorInfo()
        owner.UserId = owner_data.get("UserId", "")
        owner.SecUid = owner_data.get("SecUid", "")
        owner.Nickname = owner_data.get("Nickname", "")
        owner.HeadUrl = owner_data.get("HeadUrl", "")
        owner.FollowStatus = owner_data.get("FollowStatus", 0)
        
        return owner
    
    @staticmethod
    def parse_common_msg_fields(data: Dict[str, Any], msg: Msg) -> Msg:
        """解析通用消息字段"""
        msg.MsgId = data.get("MsgId", 0)
        msg.Content = data.get("Content", "")
        msg.RoomId = data.get("RoomId", "")
        msg.WebRoomId = data.get("WebRoomId", "")
        msg.RoomTitle = data.get("RoomTitle", "")
        msg.IsAnonymous = data.get("IsAnonymous", False)
        msg.Appid = data.get("Appid", "")
        
        # 解析用户数据
        if "User" in data and data["User"]:
            msg.User = MessageParser.parse_user(data["User"])
        
        # 解析主播数据，优先使用正确拼写的"Owner"
        if "Owner" in data and data["Owner"]:
            msg.Owner = MessageParser.parse_anchor_info(data["Owner"])
        # 如果没有"Owner"字段才考虑拼写错误的"Onwer"字段，并且确保不是字符串
        elif "Onwer" in data and data["Onwer"] and not isinstance(data["Onwer"], str):
            msg.Owner = MessageParser.parse_anchor_info(data["Onwer"])
            
        return msg
    
    @staticmethod
    def parse_danmaku(data: Dict[str, Any]) -> Msg:
        """解析弹幕消息"""
        msg = Msg()
        return MessageParser.parse_common_msg_fields(data, msg)
    
    @staticmethod
    def parse_like(data: Dict[str, Any]) -> LikeMsg:
        """解析点赞消息"""
        like_msg = LikeMsg()
        MessageParser.parse_common_msg_fields(data, like_msg)
        like_msg.Count = data.get("Count", 0)
        like_msg.Total = data.get("Total", 0)
        return like_msg
    
    @staticmethod
    def parse_member(data: Dict[str, Any]) -> MemberMessage:
        """解析进入直播间消息"""
        member_msg = MemberMessage()
        MessageParser.parse_common_msg_fields(data, member_msg)
        member_msg.CurrentCount = data.get("CurrentCount", 0)
        member_msg.EnterTipType = data.get("EnterTipType", 0)
        return member_msg
    
    @staticmethod
    def parse_gift(data: Dict[str, Any]) -> GiftMsg:
        """解析礼物消息"""
        gift_msg = GiftMsg()
        MessageParser.parse_common_msg_fields(data, gift_msg)
        gift_msg.GiftId = data.get("GiftId", 0)
        gift_msg.GiftName = data.get("GiftName", "")
        gift_msg.GroupId = data.get("GroupId", 0)
        gift_msg.GiftCount = data.get("GiftCount", 0)
        gift_msg.RepeatCount = data.get("RepeatCount", 0)
        gift_msg.DiamondCount = data.get("DiamondCount", 0)
        gift_msg.GiftValue = data.get("GiftValue", 0)  # 兼容原有代码
        gift_msg.Combo = data.get("Combo", False)
        gift_msg.ImgUrl = data.get("ImgUrl", "")
        
        # 解析接收礼物的用户
        if "ToUser" in data and data["ToUser"]:
            gift_msg.ToUser = MessageParser.parse_user(data["ToUser"])
            
        return gift_msg
    
    @staticmethod
    def parse_statistics(data: Dict[str, Any]) -> UserSeqMsg:
        """解析直播间统计消息"""
        stats_msg = UserSeqMsg()
        MessageParser.parse_common_msg_fields(data, stats_msg)
        stats_msg.OnlineUserCount = data.get("OnlineUserCount", 0)
        stats_msg.TotalUserCount = data.get("TotalUserCount", 0)
        stats_msg.TotalUserCountStr = data.get("TotalUserCountStr", "0")
        stats_msg.OnlineUserCountStr = data.get("OnlineUserCountStr", "0")
        return stats_msg
    
    @staticmethod
    def parse_fansclub(data: Dict[str, Any]) -> FansclubMsg:
        """解析粉丝团消息"""
        fansclub_msg = FansclubMsg()
        MessageParser.parse_common_msg_fields(data, fansclub_msg)
        fansclub_msg.Type = data.get("Type", 0)
        fansclub_msg.Level = data.get("Level", 0)
        fansclub_msg.FansClubName = data.get("FansClubName", "")
        return fansclub_msg
    
    @staticmethod
    def parse_share(data: Dict[str, Any]) -> ShareMessage:
        """解析直播间分享消息"""
        share_msg = ShareMessage()
        MessageParser.parse_common_msg_fields(data, share_msg)
        share_msg.ShareType = ShareType(data.get("ShareType", 0))
        return share_msg
    
    @staticmethod
    def parse_follow(data: Dict[str, Any]) -> Msg:
        """解析关注消息"""
        follow_msg = Msg()
        return MessageParser.parse_common_msg_fields(data, follow_msg)

File: DouyinBarrageGrab/Demos/NodeJS/parsers.js
const { 
    MsgUser, FansClubInfo, Msg, LikeMsg, MemberMessage, UserSeqMsg, 
    Gender, GiftMsg, FansclubMsg, ShareMessage, RoomAnchorInfo, ShareType 
} = require('./entities');

/**
 * 消息解析器类
 */
class MessageParser {
    /**
     * 解析用户数据
     * @param {Object} userData - 用户数据对象
     * @returns {MsgUser} - 解析后的用户对象
     */
    static parseUser(userData) {
        const user = new MsgUser();
        if (!userData) {
            return user;
        }
        
        user.Id = userData.Id || 0;
        user.IsAdmin = userData.IsAdmin || false;
        user.IsAnchor = userData.IsAnchor || false;
        user.ShortId = userData.ShortId || 0;
        user.DisplayId = userData.DisplayId || "";
        user.Nickname = userData.Nickname || "匿名用户";
        user.Level = userData.Level || 0;
        user.PayLevel = userData.PayLevel || 0;
        user.Gender = userData.Gender || 0;
        user.HeadImgUrl = userData.HeadImgUrl || "";
        user.SecUid = userData.SecUid || "";
        user.FollowerCount = userData.FollowerCount || 0;
        user.FollowStatus = userData.FollowStatus || 0;
        user.FollowingCount = userData.FollowingCount || 0;
        
        // 解析粉丝团信息
        if (userData.FansClub) {
            const fansClubData = userData.FansClub;
            user.FansClub = new FansClubInfo({
                ClubName: fansClubData.ClubName || "",
                Level: fansClubData.Level || 0
            });
        }
        
        return user;
    }
    
    /**
     * 解析主播信息
     * @param {Object} ownerData - 主播数据对象
     * @returns {RoomAnchorInfo|null} - 解析后的主播信息对象或null
     */
    static parseAnchorInfo(ownerData) {
        if (!ownerData) {
            return null;
        }
        
        const owner = new RoomAnchorInfo({
            UserId: ownerData.UserId || "",
            SecUid: ownerData.SecUid || "",
            Nickname: ownerData.Nickname || "",
            HeadUrl: ownerData.HeadUrl || "",
            FollowStatus: ownerData.FollowStatus || 0
        });
        
        return owner;
    }
    
    /**
     * 解析通用消息字段
     * @param {Object} data - 消息数据对象
     * @param {Msg} msg - 消息对象
     * @returns {Msg} - 解析后的消息对象
     */
    static parseCommonMsgFields(data, msg) {
        msg.MsgId = data.MsgId || 0;
        msg.Content = data.Content || "";
        msg.RoomId = data.RoomId || "";
        msg.WebRoomId = data.WebRoomId || "";
        msg.Appid = data.Appid || "";
        
        // 解析用户数据
        if (data.User) {
            msg.User = MessageParser.parseUser(data.User);
        }
        
        // 解析主播数据，优先使用正确拼写的"Owner"
        if (data.Owner) {
            msg.Owner = MessageParser.parseAnchorInfo(data.Owner);
        } 
        // 如果没有"Owner"字段才考虑拼写错误的"Onwer"字段，并且确保不是字符串
        else if (data.Onwer && typeof data.Onwer !== 'string') {
            msg.Owner = MessageParser.parseAnchorInfo(data.Onwer);
        }
        
        return msg;
    }
    
    /**
     * 解析弹幕消息
     * @param {Object} data - 弹幕消息数据
     * @returns {Msg} - 解析后的弹幕消息对象
     */
    static parseDanmaku(data) {
        const msg = new Msg();
        return MessageParser.parseCommonMsgFields(data, msg);
    }
    
    /**
     * 解析点赞消息
     * @param {Object} data - 点赞消息数据
     * @returns {LikeMsg} - 解析后的点赞消息对象
     */
    static parseLike(data) {
        const likeMsg = new LikeMsg();
        MessageParser.parseCommonMsgFields(data, likeMsg);
        likeMsg.Count = data.Count || 0;
        likeMsg.Total = data.Total || 0;
        return likeMsg;
    }
    
    /**
     * 解析进入直播间消息
     * @param {Object} data - 进入直播间消息数据
     * @returns {MemberMessage} - 解析后的进入直播间消息对象
     */
    static parseMember(data) {
        const memberMsg = new MemberMessage();
        MessageParser.parseCommonMsgFields(data, memberMsg);
        memberMsg.CurrentCount = data.CurrentCount || 0;
        memberMsg.EnterTipType = data.EnterTipType || 0;
        return memberMsg;
    }
    
    /**
     * 解析礼物消息
     * @param {Object} data - 礼物消息数据
     * @returns {GiftMsg} - 解析后的礼物消息对象
     */
    static parseGift(data) {
        const giftMsg = new GiftMsg();
        MessageParser.parseCommonMsgFields(data, giftMsg);
        giftMsg.GiftId = data.GiftId || 0;
        giftMsg.GiftName = data.GiftName || "";
        giftMsg.GroupId = data.GroupId || 0;
        giftMsg.GiftCount = data.GiftCount || 0;
        giftMsg.RepeatCount = data.RepeatCount || 0;
        giftMsg.DiamondCount = data.DiamondCount || 0;
        giftMsg.GiftValue = data.GiftValue || 0;  // 兼容原有代码
        giftMsg.Combo = data.Combo || false;
        giftMsg.ImgUrl = data.ImgUrl || "";
        
        // 解析接收礼物的用户
        if (data.ToUser) {
            giftMsg.ToUser = MessageParser.parseUser(data.ToUser);
        }
        
        return giftMsg;
    }
    
    /**
     * 解析直播间统计消息
     * @param {Object} data - 直播间统计消息数据
     * @returns {UserSeqMsg} - 解析后的直播间统计消息对象
     */
    static parseStatistics(data) {
        const statsMsg = new UserSeqMsg();
        MessageParser.parseCommonMsgFields(data, statsMsg);
        statsMsg.OnlineUserCount = data.OnlineUserCount || 0;
        statsMsg.TotalUserCount = data.TotalUserCount || 0;
        statsMsg.TotalUserCountStr = data.TotalUserCountStr || "0";
        statsMsg.OnlineUserCountStr = data.OnlineUserCountStr || "0";
        return statsMsg;
    }
    
    /**
     * 解析粉丝团消息
     * @param {Object} data - 粉丝团消息数据
     * @returns {FansclubMsg} - 解析后的粉丝团消息对象
     */
    static parseFansclub(data) {
        const fansclubMsg = new FansclubMsg();
        MessageParser.parseCommonMsgFields(data, fansclubMsg);
        fansclubMsg.Type = data.Type || 0;
        fansclubMsg.Level = data.Level || 0;
        fansclubMsg.FansClubName = data.FansClubName || "";
        return fansclubMsg;
    }
    
    /**
     * 解析直播间分享消息
     * @param {Object} data - 直播间分享消息数据
     * @returns {ShareMessage} - 解析后的直播间分享消息对象
     */
    static parseShare(data) {
        const shareMsg = new ShareMessage();
        MessageParser.parseCommonMsgFields(data, shareMsg);
        shareMsg.ShareType = data.ShareType || ShareType.未知;
        return shareMsg;
    }
    
    /**
     * 解析关注消息
     * @param {Object} data - 关注消息数据
     * @returns {Msg} - 解析后的关注消息对象
     */
    static parseFollow(data) {
        const followMsg = new Msg();
        return MessageParser.parseCommonMsgFields(data, followMsg);
    }
}

module.exports = {
    MessageParser
};